pennylane/_queuing.py

```diff
@@ -110,13 +122,59 @@ class Queue(QueuingContext):
     def __init__(self):
         self.queue = []
 
-    def _append(self, obj):
+    def _append(self, obj, **kwargs):
         self.queue.append(obj)
 
     def _remove(self, obj):
         self.queue.remove(obj)
 
 
+class AnnotatedQueue(QueuingContext):
+    """Lightweight class that maintains a basic queue of operations, in addition
+    to annotations."""
+
+    def __init__(self):
+        self.queue = OrderedDict()
+
+    def _append(self, obj, **kwargs):
+        self.queue[obj] = kwargs
+
+    def _remove(self, obj):
+        del self.queue[obj]
```

Comment:
:thinking: why not an unassigned `pop`?

Reply:
> In fact, you should always use `del` to remove an item from a dictionary. `dict.pop` and `dict.popitem` are used to remove an item and return the removed item so that it can be saved for later. If you do not need to save it however, then using these methods is less efficient.

from https://stackoverflow.com/a/27155843/10958457.

Not sure how much more efficient it is, but I've always defaulted to `del` for dictionaries when I don't need the assignment!